### PR TITLE
Sign Receipt Upload endpoint

### PR DESCRIPTION
**Pull request by garyhtou**
Closes https://github.com/hackclub/bank/issues/2542 and #2559

WIP. TODO:
- only require valid signature for unauthenticated
- update mailer to use new URL with signature
- 10 day grace period for existing urls

### CODE REVIEW

**Comment by garyhtou** [comment]
Now that we're signing these urls, we can also associate the uploaded receipt (when the user's not authenticated) with the user who received the email.

For example, you click on the link (not signed in) and upload the receipt. Instead of the receipt showing as uploaded by `anonymous`, it would be you. This would include some weird edge cases when sharing the link with someone else (such as by forwarding the email). It would show that the receipt was uploaded by you if someone else uses your link.

**Comment by garyhtou** [comment]
@maxwofford, can we get another set of eyes on this to double check that its runs properly? (we have to resolve some credential conflicts)

File: app/policies/hcb_code_policy.rb
Comment (garyhtou) [inline app/policies/hcb_code_policy.rb]: Update grace period date before merging

File: app/views/canonical_pending_transaction_mailer/notify_approved.html.erb
Comment (maxwofford) [inline app/views/canonical_pending_transaction_mailer/notify_approved.html.erb]: Could we drop this into a view helper or class method? I don't see many examples of calling services directly in our view layer.

File: app/controllers/hcb_codes_controller.rb
Comment (maxwofford) [inline app/controllers/hcb_codes_controller.rb]: Great that we attempt to handle this, love it!

**Review by maxwofford — APPROVED** [review]
Looks great! Left a couple comments, but if this is functional I'm fine with it being merged.

**Review by maxwofford — COMMENTED** [review]
Hang tight– just realized I hadn't fetched. Looks like there's some rebasing issues with the credentials when pulling in main. Could you merge in the latest from `main` & lmk once it's ready to be tested?

File: app/views/canonical_pending_transaction_mailer/notify_approved.html.erb
Comment (garyhtou) [reply app/views/canonical_pending_transaction_mailer/notify_approved.html.erb]: Would it make more sense to pass it via the mailer call?

Comment (maxwofford) [reply app/views/canonical_pending_transaction_mailer/notify_approved.html.erb]: That'd work too!

**Comment by exu3** [comment]
It seems to be up to date with `main` & works locally for me. It should be ready to be tested.

**Comment by maxwofford** [comment]
Testing locally works for me– played with the dates too and everything seems to be working. LGTM!

**Pull request merged**
